src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -500,10 +503,63 @@ export class SearchSource {
     }
   }
 
+  private checkMatch(sourceFilters: string[], field: string) {
+    return sourceFilters.some((sourceFilter) => field.match(sourceFilter));
+  }
+
+  private getFieldsWithoutSourceFilters(
+    index: IndexPattern | undefined,
+    bodyFields: SearchFieldValue[]
+  ) {
+    if (!index) {
+      return bodyFields;
+    }
+    const { fields } = index;
+    const sourceFilters = index.getSourceFiltering();
+    if (!sourceFilters || sourceFilters.excludes?.length === 0 || bodyFields.length === 0) {
+      return bodyFields;
+    }
+    const sourceFiltersValues = sourceFilters.excludes;
+    const wildcardField = bodyFields.find(
+      (el: SearchFieldValue) => el === '*' || (el as Record<string, string>).field === '*'
+    );
+    if (!wildcardField) {
+      // we already have an explicit list of fields, so we just remove source filters from that list
+      return bodyFields.filter((fld: SearchFieldValue) => {
+        if (typeof fld === 'string') {
+          return !this.checkMatch(sourceFiltersValues, fld) && !META_FIELDS.includes(fld);
+        }
+        if (!fld.hasOwnProperty('field')) {
+          return false;
+        }
+        const searchFieldValue = fld as Record<string, string>;
+        return (
+          !this.checkMatch(sourceFiltersValues, searchFieldValue.field) &&
+          !META_FIELDS.includes(searchFieldValue.field)
+        );
+      });
+    }
+    // we need to get the list of fields from an index pattern
+    const fieldsToInclude = fields.filter((field: IndexPatternField) => {
+      return !this.checkMatch(sourceFiltersValues, field.name) && !META_FIELDS.includes(field.name);
```

Comment:
It seems this filter is basically the same as the one you have above, and each place you call `checkMatch` you are also doing a check against `META_FIELDS`.

In `flatten` we also have a little helper function that can reduce the complexity of some of these conditionals by extracting the field name (L578):
```ts
    const getFieldName = (fld: string | Record<string, any>): string =>
      typeof fld === 'string' ? fld : fld.field;
```

I wonder if you might instead make a separate filter function that can be reused in both places, and also take advantage of the field name helper function to remove some unnecessary conditionals, then L526-L557 could be simplified to something like:

```ts
const getFieldName = (fld: string | Record<string, any>): string =>
  typeof fld === 'string' ? fld : fld.field;

const filterSourceFields = (fieldName: string) => {
  return (
    fieldName &&
    !sourceFiltersValues.some((sourceFilter) => fieldName.match(sourceFilter)) &&
    !getConfig(UI_SETTINGS.META_FIELDS).includes(fieldName)
  );
};

if (!wildcardField) {
  return bodyFields.filter((fld: SearchFieldValue) => filterSourceFields(getFieldName(fld)));
}

return fields.filter((fld: IndexPatternField) => filterSourceFields(fld.name))
  .map((fld: IndexPatternField) => ({
    field: fld.name,
    ...(wildcardField?.include_unmapped && { include_unmapped: wildcardField.include_unmapped }),
  }));
```